File: android/app/src/main/java/com/typeeasy/KeyboardSuggestionMatcher.kt
package com.typeeasy

object KeyboardSuggestionMatcher {
    fun matchScore(query: String, word: String): Int? {
        if (word.startsWith(query)) {
            return if (query.length == word.length) 0 else 1
        }
        if (query.length < 2 || word.length < query.length) return null
        val head = word.take(minOf(word.length, query.length + 3))
        val dist = levenshtein(query, head.take(query.length))
        if (dist <= 1) return 2 + dist
        if (fuzzySubsequencePrefix(query, head)) return 4
        return null
    }

    fun displayForm(word: String, query: String): String {
        if (query.isEmpty()) return word
        return if (query.first().isUpperCase()) {
            word.replaceFirstChar { if (it.isLowerCase()) it.titlecase() else it.toString() }
        } else {
            word.lowercase()
        }
    }

    private fun fuzzySubsequencePrefix(query: String, head: String): Boolean {
        if (query.length < 2 || head.length < query.length) return false
        var qi = 0
        for (ch in head) {
            if (qi >= query.length) break
            if (ch == query[qi]) qi++
        }
        return qi == query.length && head.length <= query.length + 6
    }

    private fun levenshtein(a: String, b: String): Int {
        if (a.isEmpty()) return b.length
        if (b.isEmpty()) return a.length
        val prev = IntArray(b.length + 1) { it }
        val curr = IntArray(b.length + 1)
        for (i in a.indices) {
            curr[0] = i + 1
            for (j in b.indices) {
                val cost = if (a[i] == b[j]) 0 else 1
                curr[j + 1] = minOf(prev[j + 1] + 1, curr[j] + 1, prev[j] + cost)
            }
            for (k in prev.indices) prev[k] = curr[k]
        }
        return prev[b.length]
    }
}

File: ios/Shared/KeyboardSuggestionMatcher.swift
import Foundation

/// Shared fuzzy prefix matching for keyboard suggestion sources.
enum KeyboardSuggestionMatcher {
  /// Lower score = better match. `nil` = no match.
  static func matchScore(query: String, word: String) -> Int? {
    if word.hasPrefix(query) {
      return query.count == word.count ? 0 : 1
    }
    guard query.count >= 2, word.count >= query.count else { return nil }

    let head = String(word.prefix(min(word.count, query.count + 3)))
    let dist = levenshtein(query, String(head.prefix(query.count)))
    if dist <= 1 { return 2 + dist }

    if fuzzySubsequencePrefix(query, in: head) {
      return 4
    }
    return nil
  }

  static func displayForm(_ word: String, query: String) -> String {
    guard let first = query.first else { return word }
    if query.allSatisfy({ $0.isUppercase }) {
      return word.uppercased()
    }
    if first.isUppercase {
      return word.prefix(1).uppercased() + word.dropFirst()
    }
    return word
  }

  private static func fuzzySubsequencePrefix(_ query: String, in head: String) -> Bool {
    guard query.count >= 2, head.count >= query.count else { return false }
    var qi = query.startIndex
    for ch in head {
      if qi == query.endIndex { break }
      if ch == query[qi] {
        qi = query.index(after: qi)
      }
    }
    return qi == query.endIndex && head.count <= query.count + 6
  }

  private static func levenshtein(_ a: String, _ b: String) -> Int {
    let aChars = Array(a)
    let bChars = Array(b)
    if aChars.isEmpty { return bChars.count }
    if bChars.isEmpty { return aChars.count }
    var prev = Array(0...bChars.count)
    var curr = [Int](repeating: 0, count: bChars.count + 1)
    for i in 1...aChars.count {
      curr[0] = i
      for j in 1...bChars.count {
        let cost = aChars[i - 1] == bChars[j - 1] ? 0 : 1
        curr[j] = min(prev[j] + 1, curr[j - 1] + 1, prev[j - 1] + cost)
      }
      swap(&prev, &curr)
    }
    return prev[bChars.count]
  }
}
